File: Multivariate/chapter2.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

heptathlon = pd.read_csv("D:/knou/Multivariate/data/heptathlon.csv")
#print(heptathlon.head())
#print(heptathlon.columns)

#기술통계량 구하기 - 소수점 이하 2자리까지 반올림
round(heptathlon.describe(), 2)

# 변환 : 변수최대값 - 변수값
heptathlon.hurdles = np.max(heptathlon.hurdles) - heptathlon.hurdles
heptathlon.run200m = np.max(heptathlon.run200m) - heptathlon.run200m
heptathlon.run800m = np.max(heptathlon.run800m) - heptathlon.run800m

#print(heptathlon.head())

# 분석변수 선택하기 
feature = ['hurdles', 'highjump', 'shot', 'run200m', 'longjump', 'javelin', 'run800m']
hep_data = heptathlon[feature]


# 파이썬에서 PCA 분석을 하기 전에 데이터를 표준화(Z-score 정규화)
# StandardScaler()는 각 열(변수)을 평균 0, 표준 편차 1로 변환하는 역할
x = StandardScaler().fit_transform(hep_data)

# 주성분분석
# PCA : 사이킷런(Scikit-learn)에서 제공하는 주성분 분석(PCA) 함수
# 주성분 개수를 hep_data의 전체 열 개수로 설정
pca_init = PCA(n_components=len(hep_data.columns))
pca_init.fit(x)


# 스크리 그림 그리기 
# explained_variance_는 각 주성분(PC)이 설명하는 실제 분산 크기를 반환
# '-o' :  각 주성분의 설명된 분산 크기를 선과 점으로 시각화
plt.title("scree plot")
plt.xlabel("component")
plt.ylabel("explained variance")
plt.plot(pca_init.explained_variance_, '-o') 
plt.show()

# 파이썬 주성분분석

# 주성분의 수를 2개로 하여 주성분 실행
pca = PCA(n_components=2)
# pca 속성 확인
print(dir(pca))
# fit_transform : 원래 데이터를 PCA를 이용해 새로운 좌표(주성분 점수)로 변환하는 과정
hep_pca = pca.fit_transform(x)


#주성분분석
# 데이터의 변동성을 나타냄 => 값이 높을수록 설명력이 높음
# [4.64611996 1.24408391]
print(pca.explained_variance_)

# 주성분분산 비율 
# 데이터의 변동성에서 차지하는 비율 
# [0.63718217 0.17061722] => PC1, PC2가 데이터의 약 80%를 설명
print(pca.explained_variance_ratio_)

# 주성분계수
# 각 원래 변수가 PC1, PC2를 구성하는 정도(가중치)
# PC1 : Longjump (-0.4562), hurdles (-0.4529), run200m (-0.4079)가 PC1에 크게 기여 => 주로 점프 달리기 
# PC2 : javelin (0.8417)이 매우 큰 값 → PC2는 창던지기(javelin) 특성을 주로 반영
print(np.round(pca.components_),3)


# 주성분 점수
# 각 데이터(선수)가 주성분(PC) 공간에서 어디에 위치하는지 나타냄
# 선수1 : -4.2064 → PC1(스피드 & 점프 능력)에서 낮은 점수
# 선수2 : C1에서 낮은 값 → 스피드 & 점프 능력 낮음
# 선수5 : PC2의 값이 크므로 투척능 력이 높은 선수
print(hep_pca[0:5,])

# PCA의 고유벡터는 임의적으로 결정될 수 있음. 부호는 영향을 미치지 않고 계수의 크기와 상대적 패턴이 중요

File: Multivariate/chapter2.R
#1988년 서울 올림픽 육상 여성 7종 경기 결과

# 1. 자료 가져오기
install.packages("HSAUR2")
library(HSAUR2)
data(heptathlon)
head(heptathlon)
# 요약
summary(heptathlon)
# csv로 저장하기 
write.csv(heptathlon, file="D:/knou/Multivariate/data/heptathlon.csv")

# 2. 자료변형하기 
# 멀리뛰기, 창던지기, 포환던지기와 같은 데이터는 값이 클수록 좋은 점수이다
# 그런데 hurdles, run200m, run800m은 작은 값일수록 좋은 점수이기 때문에 자료를 변형
# 최대값에서 해당하는 기록을 빼면 값이 클수록 좋은값으로 변환해 줌
heptathlon$hurdles = max(heptathlon$hurdle) - heptathlon$hurdles
heptathlon$run200m = max(heptathlon$run200m) - heptathlon$run200m
heptathlon$run800m = max(heptathlon$run800m) - heptathlon$run800m


# 3. 주성분 분석
# 8번째 열을 제거하고 나머지 열만 선택 =>score 열 제거
hep_data = heptathlon[,-8]

#princomp() : 주성분 분석(PCA)을 수행하는 함수 (R에서 사용됨)
#hep_data : PCA를 수행할 데이터 (8번째 열을 제거한 육상 선수 데이터)
#cor=T : 상관행렬을 사용하여 PCA를 수행
#기본적으로 princomp()는 공분산 행렬을 사용하지만, 각 변수가 다른 단위를 가지는 경우 cor=T로 설정하면 표준화된 값(Z-score)을 기반으로 분석하게 됩니다.
#scores=T : 각 데이터의 주성분 점수를 저장 

hep_pca = princomp(hep_data, cor=T, scores=T)
#생성한 PCA 결과 객체의 구성 요소
names(hep_pca)
#"sdev" (Standard Deviations, 주성분들의 표준 편차
#"loadings" (주성분 로딩, Principal Component Loadings) 가중치
# "center" (데이터 평균,
# "scale" (변수들의 표준 편차
# n.obs" (관측치 개수
# "scores" (주성분 점수
# call : 재현성을 위해 원래 사용한 함수 호출 내용을 저장

# 4. 주성분분석 결과 
# 1) 결과 요약
summary(hep_pca)

#표준편차 제곱하여 고유값 구하기
eig_val = hep_pca$sdev^2

#스크리 그림과 주성분 계수
screeplot(hep_pca, type="lines", pch=19, main="scree plot")

#누적분선그림 그리기 
# 각 주성분이 전체 데이터의 분산에서 차지하는 비율
hep_var = hep_pca$sdev^2
hep_var_ratio = hep_var/sum(hep_var)
round(hep_var_ratio,3) # 소수점ㅅ 3자리까지 반올림

#cumsum : 각 주성분까지의 누적 분산 비율을 계산
plot(cumsum(hep_var_ratio), type='b', pch=19, xlab="Component", ylab='Cumulative Proportion')
title('Varian Explained')

#첫 번째(PC1)와 두 번째(PC2) 주성분에 대한 로딩 계수(가중치)를 출력
# 값이 높을수록 영향을 많이 준다 마이너스는 주성분과 반대 방향으로 작용
#longjump (0.456), hurdles (0.453), run200m (0.408), run800m (0.375), highjump (0.377) 관련성이 높음 => 달리기와 점프 관련성이 높음
# javelin (-0.842): PC2에서 압도적으로 중요한 변수 => 청던지기와 밀접한 관계

hep_pca$loadings[,c(1:2)]


#3) 주성분 점수 및 행렬도 
hep_pca$scores[, c(1:2)]
# highjump, run800, hurdles, longjump가 서로 가까운데 위치 벡터 방향도 비슷 
# run200, shot이 가깝게 위치, javelin은 다른 변수들과 다른 위치
# 가까운 거리와 방향일수록 변수들의 상관성이 높다 
biplot(hep_pca, cex=0.7, col=c("Red","Blue"))
title("Biplot")


# 주성분분석2
beer = read.csv("D:/knou/Multivariate/data/beer.csv")
summary(beer)

# 상관계수행렬 및 산점도 행렬 보기 
# 가격, 크기, 알코올함량이 상관계수가 높음 
# 평판은 다른 변수들과 상관계수가 비교적 높은 않으며, 다른 변수들과 모두 음의 상관계수를 나타냈다
round(cor(beer),2)

# 산점도 행렬
# color vs. aroma: 비교적 점들이 선형적으로 증가하는 경향이 있어 보이며, 양의 상관관계를 가질 가능성이 있음
# aroma vs. taste: 비슷한 경향을 보이며, aroma가 높을수록 taste도 높아지는 양의 상관관계가 존재할 가능성이 있음
plot(beer, pch=19)

# 주성분분석 실행
# cor = F는 공분산 이용, score=F : 각 케이스의 주성분 성분을 포함하라는 의미
beer_pca = princomp(beer, cor=F, scores=F)
beer_pca

# 결과 요약
# 첫번째 주성분이 47% 두번째 37% 설명력을 가짐
summary(beer_pca)

# 스크리 그림 및 주성분 계수
screeplot(beer_pca, type="lines", pch=19)
# 스크리 그림에서 주성분을 3개로 판단하여 1:3까지 계수 확인
beer_pca$loadings[,c(1:3)]




